### PR TITLE
First cut OrderedRDD2

### PR DESCRIPTION
**Pull request by cseed**
It needs some more cleanup, I plan to make another review pass myself.  I can definitely split it into multiple pieces if necessary.  Will post latest benchmark numbers in a moment.

### CODE REVIEW

**Comment by cseed** [comment]
vs 0.1:

```
0.1-5b9d9c0	import_vcf/count	[23.49864888191223, 20.585176944732666, 19.687045097351074]
0.1-5b9d9c0	import_vcf_genotype/write	[23.210124969482422, 21.06765389442444, 20.78858995437622]
0.1-5b9d9c0	import_vcf/write	[21.205545902252197, 20.95118498802185, 21.136919021606445]
0.1-5b9d9c0	read/count	[1.2925410270690918, 0.6846837997436523, 0.7131791114807129]
0.1-5b9d9c0	import_vcf/query_genotypes	[23.002583026885986, 21.8937349319458, 21.95244598388672]
0.1-5b9d9c0	read/query_genotypes	[2.6910181045532227, 2.763874053955078, 2.8131890296936035]
206M	profile.vds

devel-4e1a92c	import_vcf/count	[13.741026878356934, 11.236669063568115, 11.742599964141846]
devel-4e1a92c	import_vcf_genotype/write	[39.8490309715271, 38.8665668964386, 38.8463978767395]
devel-4e1a92c	import_vcf/write	[22.137010097503662, 22.579964876174927, 22.09669589996338]
devel-4e1a92c	read/count	[3.0012340545654297, 2.2702929973602295, 2.221552848815918]
devel-4e1a92c	import_vcf/query_genotypes	[13.756996154785156, 14.001467943191528, 13.464909791946411]
devel-4e1a92c	read/query_genotypes	[4.777192115783691, 4.960904121398926, 4.8691558837890625]
241M	profile.vds
```

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 1]
```diff
@@ -145,10 +172,46 @@ class Decoder(in: DataInputStream) {
   }
 }
 
-class Encoder(out: DataOutputStream) {
+final class Encoder() {
+  var outMem: Array[Byte] = new Array[Byte](8 * 1024)
```
Comment (danking) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: should we have a `sizeHint` instead that defaults to 8KB?

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 2]
```diff
@@ -277,29 +342,44 @@ class RichRDDUnsafeRow(val rdd: RDD[UnsafeRow]) extends AnyVal {
...
-      sHadoopConfBc.value.value.writeLZ4DataFile(path + "/rowstore/part-" + pis,
-        64 * 1024,
-        LZ4Factory.fastestInstance().highCompressor()) { out =>
-        val en = new Encoder(out)
+      sHadoopConfBc.value.value.writeFile(path + "/rowstore/part-" + pis) { out =>
+        var comp: Array[Byte] = null
+        val en = new Encoder()
+
+        it.foreach { rv =>
+          // println("write rv", rv.pretty(t))
```
Comment (danking) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: delete

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 3]
```diff
@@ -277,29 +342,44 @@ class RichRDDUnsafeRow(val rdd: RDD[UnsafeRow]) extends AnyVal {
...
-          val rowSize = r.region.offset
-          out.writeLong(rowSize)
+          val maxCompLen = LZ4Utils.maxCompressedLength(decompLen)
+          if (comp == null || maxCompLen > comp.length)
+            comp = new Array[Byte](maxCompLen)
 
-          en.writeRegionValue(t, r.region, r.offset)
+          val compLen = LZ4Utils.compress(comp, decomp, decompLen)
+
+          // println("write", decompLen, compLen)
```
Comment (danking) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: delete

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 4]
```diff
@@ -277,29 +342,44 @@ class RichRDDUnsafeRow(val rdd: RDD[UnsafeRow]) extends AnyVal {
...
+
+          writeInt(out, decompLen)
+          writeInt(out, compLen)
+          out.write(comp, 0, compLen)
 
           rowCount += 1
         }
 
-        out.writeLong(-1)
+        // -1
```
Comment (danking) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: delete

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 5]
```diff
@@ -312,54 +392,81 @@ class RichRDDUnsafeRow(val rdd: RDD[UnsafeRow]) extends AnyVal {
...
 case class ReadRowsRDDPartition(index: Int) extends Partition
 
+object ReadRowsRDD {
+  def readInt(in: InputStream): Int = {
+    val b0 = in.read()
+    val b1 = in.read()
+    val b2 = in.read()
+    val b3 = in.read()
+
+    (b0 & 0xFF) |
```
Comment (danking) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: [`InputStream.read`](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#read--) ensures the high  bits are zero unless the end of stream is reached (well, really, it says the value will be between 0 and 255, and AFAIK, java guarantees two's complement representation). Is the `& 0xff` necessary? In the case that the stream terminates half-way through an int, no sensible value could be returned anyway, so it doesn't seem bad for the returned Int to also be `-1`.

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 6]
```diff
@@ -312,54 +392,81 @@ class RichRDDUnsafeRow(val rdd: RDD[UnsafeRow]) extends AnyVal {
...
+      def hasNext: Boolean = decompLen != -1
 
-      def next(): UnsafeRow = {
+      def next(): RegionValue = {
         if (!hasNext)
           throw new NoSuchElementException("next on empty iterator")
 
+        val compLen = ReadRowsRDD.readInt(in)
+
+        // println("read", decompLen, compLen)
```
Comment (danking) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: delete

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 7]
```diff
@@ -1,21 +1,49 @@
 package is.hail.annotations
```
Comment (danking) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: Meta-comment: I'm rather inclined not to pile several classes into one file whose name is one of those classes. Seems messy.

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 8]
```diff
@@ -312,54 +392,81 @@ class RichRDDUnsafeRow(val rdd: RDD[UnsafeRow]) extends AnyVal {
...
+          decomp = new Array[Byte](decompLen)
+        LZ4Utils.decompress(decomp, decompLen, comp, compLen)
+
+        dec.set(decomp, 0)
         region.clear()
-        region.ensure(rowSize)
+        rv.offset = dec.readRegionValue(localT.fundamentalType, region)
 
-        dec.readRegionValue(t, region)
+        // println("read rv", rv.pretty(t))
```
Comment (danking) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: delete

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 9]
```diff
@@ -30,27 +66,92 @@ class UnsafeIndexedSeqAnnotation(
 
   override def write(kryo: Kryo, output: Output) {
     kryo.writeObject(output, ttBc)
-    val enc = new Encoder(new DataOutputStream(output))
+    val enc = new Encoder()
     enc.writeRegionValue(ttBc.value.typ, region, aoff)
+    kryo.writeObject(output, enc.outOff)
+    kryo.writeObject(output, enc.outMem) // FIXME
```
Comment (danking) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: ????

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 10]
```diff
@@ -1,20 +1,56 @@
...
+  }
+}
+
+class UnsafeIndexedSeq(var region: MemoryBuffer,
   var ttBc: Broadcast[TypeTree],
-  var region: MemoryBuffer, var aoff: Long, var elemsOffset: Long,
-  var length: Int,
-  var elemSize: Long) extends IndexedSeq[Annotation] with KryoSerializable {
+  var elemSize: Long, var aoff: Long, var elemsOffset: Long,
+  var length: Int) extends IndexedSeq[Annotation] with KryoSerializable with Serializable {
```
Comment (danking) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: Making it `Serializable` could be a separate PR

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 11]
```diff
@@ -254,15 +354,50 @@ class UnsafeRow(var ttBc: Broadcast[TypeTree],
...
   override def write(kryo: Kryo, output: Output) {
     kryo.writeObject(output, ttBc)
-    val enc = new Encoder(new DataOutputStream(output))
+    val enc = new Encoder()
     enc.writeRegionValue(ttBc.value.typ, region, offset)
+    kryo.writeObject(output, enc.outOff)
+    kryo.writeObject(output, enc.outMem)
+  }
+
+  private def writeObject(out: ObjectOutputStream) {
```
Comment (danking) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: How was serialization working before? maybe we can keep this separate as well to bring down the size of this PR?

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 12]
```diff
@@ -254,15 +354,50 @@ class UnsafeRow(var ttBc: Broadcast[TypeTree],
...
+
+    out.writeInt(smallOutOff)
+    out.write(enc.outMem, 0, smallOutOff)
   }
 
   override def read(kryo: Kryo, input: Input) {
     val cTorrentBroadcast = Class.forName("org.apache.spark.broadcast.TorrentBroadcast")
     ttBc = kryo.readObject(input, cTorrentBroadcast).asInstanceOf[Broadcast[TypeTree]]
-    val dec = new Decoder(new DataInputStream(input))
+    val inOff = kryo.readObject[Long](input, classOf[Long])
```
Comment (danking) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: there's seriously no readLong in Kryo?

Comment (danking) [reply src/main/scala/is/hail/annotations/UnsafeRow.scala]: Could you call `input.readLong()`?

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 13]
```diff
@@ -254,15 +354,50 @@ class UnsafeRow(var ttBc: Broadcast[TypeTree],
 
   override def write(kryo: Kryo, output: Output) {
     kryo.writeObject(output, ttBc)
-    val enc = new Encoder(new DataOutputStream(output))
+    val enc = new Encoder()
     enc.writeRegionValue(ttBc.value.typ, region, offset)
+    kryo.writeObject(output, enc.outOff)
```
Comment (danking) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: Could we use `output.writeLong(Long)`?

File: src/main/scala/is/hail/annotations/UnsafeUtils.scala [hunk 14]
```diff
@@ -3,15 +3,7 @@ package is.hail.annotations
...
-  def arrayElementSize(t: Type): Long = {
-    var eltSize = t.byteSize
-    if (eltSize > 0) {
-      val mod = eltSize % t.alignment
-      if (mod != 0)
-        eltSize += (t.alignment - mod)
-    }
-    eltSize
-  }
+  def arrayElementSize(t: Type): Long = roundUpAlignment(t.byteSize, t.alignment)
```
Comment (danking) [inline src/main/scala/is/hail/annotations/UnsafeUtils.scala]: seems like an independent change?

File: src/main/scala/is/hail/expr/Relational.scala [hunk 15]
```diff
@@ -30,10 +30,27 @@ case class MatrixType(
...
-    "v" -> vType,
-    "va" -> vaType,
-    "gs" -> TArray(genotypeType))
+  def partitionKeyType: Type =
+    vType match {
+      case t: TVariant =>
+        TLocus(vType.asInstanceOf[TVariant].gr)
+      case _ =>
+        vType
+    }
```
Comment (danking) [inline src/main/scala/is/hail/expr/Relational.scala]: This pattern also appears in VariantSampleMatrix. I'm not sure there's a natural way to avoid this duplication...

File: src/main/scala/is/hail/expr/Relational.scala [hunk 16]
```diff
@@ -3,9 +3,9 @@ package is.hail.expr
 import java.io.FileNotFoundException
```
Comment (danking) [inline src/main/scala/is/hail/expr/Relational.scala]: Is it possible for these changes to be a separate PR depending on the changes that add `OrderedRDD2`?

File: src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 17]
```diff
@@ -1,23 +1,221 @@
 package is.hail.io.vcf
```
Comment (danking) [inline src/main/scala/is/hail/io/vcf/LoadVCF.scala]: These changes seem to just use `RegionValueBuilder`. Is this an orthogonal set of changes we could make a separate PR?

File: src/main/scala/is/hail/keytable/KeyTable.scala [hunk 18]
```diff
@@ -340,37 +339,29 @@ case class KeyTable(hc: HailContext, rdd: RDD[Row],
     copy(key = key.toArray)
   }
 
-  def select(fieldsSelect: Array[String], newKeys: Array[String]): KeyTable = {
```
Comment (danking) [inline src/main/scala/is/hail/keytable/KeyTable.scala]: All these changes seem independent of the `Unsafe`-related stuff, right?

File: src/main/scala/is/hail/sparkextras/OrderedRDD2.scala [hunk 19]
```diff
@@ -0,0 +1,733 @@
...
+import is.hail.annotations._
+import is.hail.expr.{JSONAnnotationImpex, Parser, TArray, TSet, TStruct, Type}
+import is.hail.utils._
+import org.apache.spark.{Partition, Partitioner, SparkContext, TaskContext}
+import org.apache.spark.rdd.{RDD, ShuffledRDD}
+import org.json4s.JsonAST.{JInt, JObject, JString, JValue}
+
+import scala.collection.mutable
+
+object BinarySearch {
```
Comment (danking) [inline src/main/scala/is/hail/sparkextras/OrderedRDD2.scala]: should be in `is.hail.util` with tests?

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 20]
```diff
@@ -929,12 +928,8 @@ class GenericGenotype(val _unboxedGT: Int,
 
   if (_isLinearScale) {
     require(_unboxedGQ == -1)
-    if (_unboxedPX == null)
-      require(_unboxedGT == -1)
-    else {
+    if (_unboxedPX != null)
       require(_unboxedPX.sum == 32768)
-      require(_unboxedGT == Genotype.gtFromLinear(_unboxedPX).getOrElse(-1))
-    }
```
Comment (danking) [inline src/main/scala/is/hail/variant/Genotype.scala]: this fixes a bug I saw, right? This is a totally separate issue from anything else here.

File: src/main/scala/is/hail/variant/HTSGenotypeView.scala [hunk 21]
```diff
@@ -6,7 +6,7 @@ import is.hail.utils._
 
 object HTSGenotypeView {
   def apply(rowSignature: TStruct): HTSGenotypeView = {
-    rowSignature.fields(2).typ.asInstanceOf[TArray].elementType match {
+    rowSignature.fields(3).typ.asInstanceOf[TArray].elementType match {
```
Comment (danking) [inline src/main/scala/is/hail/variant/HTSGenotypeView.scala]: I'm getting tired, but why did this change?

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 22]
```diff
@@ -843,14 +817,6 @@ class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata:
...
   }
 
-  def deleteGlobal(args: String*): (Type, Deleter) = deleteGlobal(args.toList)
-
-  def deleteGlobal(path: List[String]): (Type, Deleter) = globalSignature.delete(path)
-
-  def deleteSA(args: String*): (Type, Deleter) = deleteSA(args.toList)
-
-  def deleteSA(path: List[String]): (Type, Deleter) = saSignature.delete(path)
-
```
Comment (danking) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: I'm a big fan of code removal, but a lot of this seems independent of the rest of the PR.

File: src/test/scala/is/hail/methods/AnnotateSuite.scala [hunk 23]
```diff
@@ -645,7 +644,7 @@ class AnnotateSuite extends SparkSuite {
     assert(gen6.dropSamples().same(drop6))
 
 
-    val skt = vds.samplesKT().select(Array("s"), Array("s"))
+    val skt = vds.samplesKT().select(Array("s"))
```
Comment (danking) [inline src/test/scala/is/hail/methods/AnnotateSuite.scala]: while we're at it, get rid of `Array`

File: src/test/scala/is/hail/methods/KeyTableSuite.scala [hunk 24]
```diff
@@ -284,21 +284,21 @@ class KeyTableSuite extends SparkSuite {
...
-    assert((drop4.key sameElements Array.empty[String]) && (drop4.fieldNames sameElements Array("field1")))
+    assert((drop4.key sameElements Array.empty[String]) && (drop4.columns sameElements Array("field1")))
     val drop5 = kt.drop(Array("Sample", "field1", "field2"))
-    assert((drop5.key sameElements Array.empty[String]) && (drop5.fieldNames sameElements Array.empty[String]))
+    assert((drop5.key sameElements Array.empty[String]) && (drop5.columns sameElements Array.empty[String]))
 
     val kt2 = KeyTable(hc, rdd, signature, Array("field1", "field2"))
     val drop6 = kt2.drop(Array("field1"))
-    assert((drop6.key sameElements Array("field2")) && (drop6.fieldNames sameElements Array("Sample", "field2")))
+    assert((drop6.key sameElements Array("field2")) && (drop6.columns sameElements Array("Sample", "field2")))
```
Comment (danking) [inline src/test/scala/is/hail/methods/KeyTableSuite.scala]: renames => separate PR?

File: src/test/scala/is/hail/methods/KeyTableSuite.scala [hunk 25]
```diff
@@ -337,7 +337,7 @@ class KeyTableSuite extends SparkSuite {
       .variantsKT()
       .expandTypes()
       .flatten()
-      .select(Array("va.info.MQRankSum"), Array.empty[String])
+      .select(Array("va.info.MQRankSum"))
```
Comment (danking) [inline src/test/scala/is/hail/methods/KeyTableSuite.scala]: no need for `Array` on the keys.

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 26]
```diff
@@ -264,7 +322,14 @@ class Encoder(out: DataOutputStream) {
   }
 }
 
-class RichRDDUnsafeRow(val rdd: RDD[UnsafeRow]) extends AnyVal {
+class RichRDDRegionValue(val rdd: RDD[RegionValue]) extends AnyVal {
+  def writeInt(out: OutputStream, i: Int) {
+    out.write(i & 0xFF)
+    out.write((i >> 8) & 0xFF)
+    out.write((i >> 16) & 0xFF)
+    out.write((i >> 24) & 0xFF)
```
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: In this case, the & is unnecessary, right? We should leave it, I just want to make sure I understand what's going on.

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 27]
```diff
@@ -1,20 +1,56 @@
...
 import org.apache.spark.broadcast.Broadcast
 import org.apache.spark.sql.Row
 
-class UnsafeIndexedSeqAnnotation(
+object UnsafeIndexedSeq {
+  def apply(sc: SparkContext, region: MemoryBuffer, t: TArray, aoff: Long, length: Int): UnsafeIndexedSeq = {
+    new UnsafeIndexedSeq(region, BroadcastTypeTree(sc, t), t.elementByteSize, aoff, aoff + t.elementsOffset(length), length)
+  }
+
+  def apply(ttBc: Broadcast[TypeTree], region: MemoryBuffer, aoff: Long): UnsafeIndexedSeq = {
```
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: I think we should rip out the typetree stuff. It's unnecessary complexity to hack performance in a way that won't matter when the unsafe system is complete

File: src/main/scala/is/hail/annotations/UnsafeRow.scala (hunk 9 above)
Comment (tpoterba) [reply src/main/scala/is/hail/annotations/UnsafeRow.scala]: It's copying the full memory for each unsafeindexedSeq, right?

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 28]
```diff
@@ -30,27 +66,92 @@ class UnsafeIndexedSeqAnnotation(
 
   override def write(kryo: Kryo, output: Output) {
     kryo.writeObject(output, ttBc)
-    val enc = new Encoder(new DataOutputStream(output))
+    val enc = new Encoder()
     enc.writeRegionValue(ttBc.value.typ, region, aoff)
+    kryo.writeObject(output, enc.outOff)
```
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: `output.writeLong(enc.outoff)`

File: src/main/scala/is/hail/annotations/UnsafeRow.scala (hunk 12 above)
Comment (tpoterba) [reply src/main/scala/is/hail/annotations/UnsafeRow.scala]: I think it's because he used writeObject above, which will write a java.lang.Long. If he switches to output.writeLong, then he can use input.readLong

File: src/main/scala/is/hail/variant/HTSGenotypeView.scala (hunk 21 above)
Comment (tpoterba) [reply src/main/scala/is/hail/variant/HTSGenotypeView.scala]: now it's PK, K, Annotations, GS

**Comment by cseed** [comment]
Thanks for the comments, everyone.  Closing to address the comments and break this into multiple pieces.
